substrate/frame/asset-conversion/src/lib.rs

```diff
@@ -1018,48 +1018,44 @@ pub mod pallet {
 
 		/// Used by the RPC service to provide current prices.
 		pub fn quote_price_exact_tokens_for_tokens(
-			asset1: T::MultiAssetId,
-			asset2: T::MultiAssetId,
-			amount: T::AssetBalance,
+			path: BoundedVec<T::MultiAssetId, T::MaxSwapPathLength>,
+			amount_in: T::AssetBalance,
 			include_fee: bool,
 		) -> Option<T::AssetBalance> {
-			let pool_id = Self::get_pool_id(asset1.clone(), asset2.clone());
-			let pool_account = Self::get_pool_account(&pool_id);
-
-			let balance1 = Self::get_balance(&pool_account, &asset1).ok()?;
-			let balance2 = Self::get_balance(&pool_account, &asset2).ok()?;
-			if !balance1.is_zero() {
-				if include_fee {
-					Self::get_amount_out(&amount, &balance1, &balance2).ok()
-				} else {
-					Self::quote(&amount, &balance1, &balance2).ok()
-				}
+			Self::validate_swap_path(&path).ok()?;
+			Some(if include_fee {
+				*Self::get_amounts_out(&amount_in, &path).ok()?.last()?
 			} else {
-				None
-			}
+				let mut result = amount_in;
+				for assets_pair in path.windows(2).rev() {
+					if let [asset1, asset2] = assets_pair {
+						let (reserve_in, reserve_out) = Self::get_reserves(asset1, asset2).ok()?;
+						result = reserve_out * result / reserve_in;
```

Comment:
Checked mul/div? Is it ever possible that `reserve_in` is zero?

Reply:
Btw we have `Self::quote()` for that, which uses the internal `mul_div()` under the hood

Reply:
> Is it ever possible that reserve_in is zero?

get_reserves() will throw an error if any of the balances is zero

Reply:
Agreed using `Self::quote` is nicer. Have updated.